File: turn_offset_controller/turn_offset_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from autoware_auto_vehicle_msgs.msg import TurnIndicatorsReport, SteeringReport
from tier4_planning_msgs.msg import LateralOffset

class TurnOffsetController(Node):

    def __init__(self):
        super().__init__('turn_report_subscriber')
        self.subscription = self.create_subscription(
            TurnIndicatorsReport,
            '/vehicle/status/turn_indicators_status',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning
        
        self.steering_status_subscription = self.create_subscription(  # ステアリング状態のサブスクリプションを作成
            SteeringReport,
            '/vehicle/status/steering_status',
            self.steering_status_callback,
            10)        

        self.publisher_ = self.create_publisher(LateralOffset, '/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/input/lateral_offset', 10)

        self.prev_state = 1
        self.in_intersection = False  # 交差点内であるかを保存する変数
        
    def listener_callback(self, msg):
        self.get_logger().debug('I heard: "%d"' % msg.report)

        offset_msg = LateralOffset()

        if msg.report == self.prev_state:
            return

        if msg.report == 1 or self.in_intersection:  # 交差点にいる場合も横方向オフセットをゼロに設定
            offset_msg.lateral_offset = 0.0
        elif msg.report == 2:
            offset_msg.lateral_offset = 0.3
        elif msg.report == 3:
            offset_msg.lateral_offset = -0.3

        self.prev_state = msg.report

        self.publisher_.publish(offset_msg)
        self.get_logger().info('turn_signal_report: "%d"' % msg.report)
        self.get_logger().info('publishing_offset: "%f"' % offset_msg.lateral_offset)

    def steering_status_callback(self, msg):
        self.get_logger().debug('Steering tire angle: "%f"' % msg.steering_tire_angle)
        self.new_intersection_status = abs(msg.steering_tire_angle) >= 0.04  # ステアリング角度が2.0度以上であるかどうか
        if( self.new_intersection_status != self.in_intersection ):
            self.in_intersection = self.new_intersection_status
            if self.in_intersection:
                offset_msg = LateralOffset()
                offset_msg.lateral_offset = 0.0
                self.publisher_.publish(offset_msg)
                self.get_logger().info('In intersection, setting offset to 0.0')
            
def main(args=None):
    rclpy.init(args=args)

    my_subscriber = TurnOffsetController()

    rclpy.spin(my_subscriber)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    my_subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/turn_offset_controller.cpp
#include <rclcpp/rclcpp.hpp>
#include <autoware_auto_vehicle_msgs/msg/turn_indicators_report.hpp>
#include <autoware_auto_vehicle_msgs/msg/steering_report.hpp>
#include <tier4_planning_msgs/msg/lateral_offset.hpp>


class TurnOffsetController : public rclcpp::Node {
public:
    TurnOffsetController() : Node("turn_report_subscriber"), in_intersection_(false) {
        subscription_ = this->create_subscription<autoware_auto_vehicle_msgs::msg::TurnIndicatorsReport>(
                "/vehicle/status/turn_indicators_status", 10,
                std::bind(&TurnOffsetController::listener_callback, this, std::placeholders::_1));

        steering_status_subscription_ = this->create_subscription<autoware_auto_vehicle_msgs::msg::SteeringReport>(
                "/vehicle/status/steering_status", 10,
                std::bind(&TurnOffsetController::steering_status_callback, this, std::placeholders::_1));

        publisher_ = this->create_publisher<tier4_planning_msgs::msg::LateralOffset>(
                "/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/input/lateral_offset", 10);

        prev_state_ = 1;
        in_intersection_ = false;
    }

private:
    void listener_callback(const autoware_auto_vehicle_msgs::msg::TurnIndicatorsReport::SharedPtr msg) {
        RCLCPP_DEBUG(this->get_logger(), "I heard: %d", msg->report);

        auto offset_msg = std::make_shared<tier4_planning_msgs::msg::LateralOffset>();

        if (msg->report == prev_state_) {
            return;
        }

        if (msg->report == 1 || in_intersection_) {
            offset_msg->lateral_offset = 0.0;
        } else if (msg->report == 2) {
            offset_msg->lateral_offset = 0.3;
        } else if (msg->report == 3) {
            offset_msg->lateral_offset = -0.3;
        }

        prev_state_ = msg->report;

        publisher_->publish(*offset_msg);
        RCLCPP_INFO(this->get_logger(), "turn_signal_report: %d", msg->report);
        RCLCPP_INFO(this->get_logger(), "publishing_offset: %f", offset_msg->lateral_offset);
    }


void steering_status_callback(const autoware_auto_vehicle_msgs::msg::SteeringReport::SharedPtr msg_ptr) {
    const auto& msg = *msg_ptr; // Convert to a reference
    RCLCPP_DEBUG(this->get_logger(), "Steering tire angle: %f", msg.steering_tire_angle);

    bool new_intersection_status = std::abs(msg.steering_tire_angle) >= 0.00872664;  // 0.5[deg]
    if (new_intersection_status != in_intersection_) {
        in_intersection_ = new_intersection_status;
        if (in_intersection_) {
            auto offset_msg = std::make_shared<tier4_planning_msgs::msg::LateralOffset>();
            offset_msg->lateral_offset = 0.0;
            publisher_->publish(*offset_msg);
            RCLCPP_INFO(this->get_logger(), "In intersection, setting offset to 0.0");
        }
    }
}

    rclcpp::Subscription<autoware_auto_vehicle_msgs::msg::TurnIndicatorsReport>::SharedPtr subscription_;
    rclcpp::Subscription<autoware_auto_vehicle_msgs::msg::SteeringReport>::SharedPtr steering_status_subscription_;
    rclcpp::Publisher<tier4_planning_msgs::msg::LateralOffset>::SharedPtr publisher_;
    int prev_state_;
    bool in_intersection_;
};


int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);

    auto my_subscriber = std::make_shared<TurnOffsetController>();

    rclcpp::spin(my_subscriber);
    rclcpp::shutdown();
    return 0;
}
